torchdata/nodes/pin_memory.py

```diff
@@ -0,0 +1,163 @@
+        except queue.Empty:
+            return
+        if not done_event.is_set() and not isinstance(data, ExceptionWrapper) and not isinstance(data, StopIteration):
+            try:
+                data = pin_memory(data, device)
+            except Exception:
+                data = ExceptionWrapper(where=f"in pin memory thread for device {device_id}")
+        while not done_event.is_set():
+            try:
+                out_queue.put(data, timeout=MP_STATUS_CHECK_INTERVAL)
+                break
+            except queue.Full:
+                continue
+
+    # See NOTE [ Data Loader Multiprocessing Shutdown Logic ] for details on the
+    # logic of this function.
+    while not done_event.is_set():
+        # Make sure that we don't preserve any object from one iteration
+        # to the next
+        do_one_step()
+
+
+class PinMemory(BaseNode[T]):
+    def __init__(
+        self,
+        source: BaseNode[T],
+        pin_memory_device: str = "",
+    ):
+        self.source = source
+
+        self.in_q: queue.Queue = queue.Queue()
+        self.out_q: queue.Queue = queue.Queue()
+        self.sem = threading.BoundedSemaphore(value=1)
+
+        self._started = False
+        self._populate_queue_stop_event = threading.Event()
+        self._pin_memory_stop_event = threading.Event()
+
+        self._pin_memory = torch.cuda.is_available()
+        if len(pin_memory_device) == 0:
+            self._pin_memory_device = None
+        else:
+            self._pin_memory_device = pin_memory_device
+
+        if self._pin_memory_device == "xpu":
+            self.current_device = torch.xpu.current_device()  # type: ignore[attr-defined]
+        elif self._pin_memory_device == torch._C._get_privateuse1_backend_name():
+            custom_device_mod = getattr(torch, torch._C._get_privateuse1_backend_name())
+            self.current_device = custom_device_mod.current_device()
+        else:
+            self.current_device = torch.cuda.current_device()  # choose cuda for default
+
+        self.read_thread: Optional[threading.Thread] = None
+        self.pin_memory_thread: Optional[threading.Thread] = None
+
+    def iterator(self) -> Iterator[T]:
+        if not self._started:
+            self._populate_queue_stop_event.clear()
+            self._pin_memory_stop_event.clear()
+
+            self.read_thread = threading.Thread(
+                target=_populate_queue,
+                args=(
+                    self.source,
+                    self.in_q,
+                    self._populate_queue_stop_event,
+                    self.sem,
+                ),
+            )
+            self.pin_memory_thread = threading.Thread(
+                target=_pin_memory_loop,
+                args=(
+                    self.in_q,
+                    self.out_q,
+                    self.current_device,
+                    self._pin_memory_stop_event,
+                    self._pin_memory_device,
+                ),
+            )
```

Comment:
Would it be possible to eliminate the `read_thread` and do all reading and pinning in the `pin_memory_thread`? My thinking here is that every thread we introduce is going to introduce potential copying overhead.

My understanding of the purpose of `read_thread` is just to get items from `self.source` into `self.in_q`. In `pin_memory_thread`, we then read items from `self.in_q`, pin them, and put them in `self.out_q`. I think we could simplify things by eliminating `read_thread` and `self.in_q`. Then data would go directly from `self.source` to `self.out_q` inside `pin_memory_thread`.

Another way to think about this: we want to introduce threads for operations that are likely to be expensive. If _both_ reading and pinning are likely to be expensive, then it makes sense to use a thread for both. But my intuition is that will be faster to combine them into one thread.

Reply:
Looking at `ParallelMapper`, I think my comment above applies there as well.

Reply:
I think you're right on this one for PinMemory, since it's single-threaded. For ParallelMapper I think we want some way to guarantee single-threaded access to source, in case source is either not-threadsafe, or not properly sharded in the case of multiprocess workers. LMK if that doesn't make sense, can discuss more tomorrow